Source/WebCore/dom/FindRevealAlgorithms.cpp

```diff
@@ -37,49 +37,66 @@
 
 namespace WebCore {
 
-// https://html.spec.whatwg.org/#ancestor-details-revealing-algorithm
-void revealClosedDetailsAncestors(Node& node)
+enum class RevealType : bool {
+    ClosedDetails,
+    HiddenUntilFound
+};
+
+// https://html.spec.whatwg.org/#ancestor-revealing-algorithm
+void revealClosedDetailsAndHiddenUntilFoundAncestors(Node& node)
 {
-    if (!node.document().settings().detailsAutoExpandEnabled())
+    node.protectedDocument()->updateStyleIfNeeded();
+
+    // Bail out if there is neither a hidden=until-found or details ancestor.
+    if (node.renderStyle() && !node.renderStyle()->autoRevealsWhenFound())
         return;
 
-    Ref currentNode = node;
-    while (RefPtr parent = currentNode->parentInComposedTree()) {
-        if (RefPtr slot = currentNode->assignedSlot(); slot && slot->userAgentPart() == UserAgentParts::detailsContent() && slot->shadowHost()) {
-            currentNode = *slot->shadowHost();
-            Ref details = downcast<HTMLDetailsElement>(currentNode);
+    auto closedDetailsElementAncestor = [](Node& node) -> RefPtr<HTMLDetailsElement> {
+        RefPtr slot = node.assignedSlot();
+        if (slot && slot->userAgentPart() == UserAgentParts::detailsContent() && slot->shadowHost()) {
+            Ref details = downcast<HTMLDetailsElement>(*slot->shadowHost());
             if (!details->hasAttributeWithoutSynchronization(HTMLNames::openAttr))
-                details->toggleOpen();
-        } else
-            currentNode = *parent;
-    }
-}
+                return details;
+        }
+        return nullptr;
+    };
 
-// https://html.spec.whatwg.org/#ancestor-hidden-until-found-revealing-algorithm
-void revealHiddenUntilFoundAncestors(Node& node)
-{
-    if (!node.document().settings().hiddenUntilFoundEnabled())
-        return;
+    Vector<std::pair<Ref<HTMLElement>, RevealType>> ancestors;
+
+    RefPtr ancestor = node;
+    while (ancestor->parentElementInComposedTree()) {
```

Comment:
Might make the loop structure clearer to write it as a `for` instead of `while`.